File: VirtualLab_MatLab/TokaPlot_MATLAB/functions/TokaPlot.m
classdef TokaPlot

    properties

        field % plasma field to plot

        fig % figure

        config % a structure containing plot information and preferences

        equi % structure

        diag % diagnostic to plot

        meas % measurement to plot

    end

    methods

        %% Fields Plotting

        function fig = PlotField(obj,equi,field,fig,config)


            if nargin < 2
                disp("missing input")
                return
            elseif nargin == 2
                field = "ne";
                fig = figure();
                config = struct;
            elseif nargin == 3
                fig = figure();
                config = struct;
            elseif nargin == 4
                config = struct;
            end

            %% Get the subplot information and set if not specified

            if isfield(config, "subplot")==0
                config.subplot = [1 1 1];
            end

            %% Plot

            figure(fig)
            subplot(config.subplot(1),config.subplot(2), config.subplot(3))
           
            % complete the plot
            if isfield(config, "hold") == 1 && config.hold == "on"
                hold on
            else
                hold off
            end

            contourf(equi.geo.grid.Rg,equi.geo.grid.Zg,equi.(field).*equi.geo.wall.inside, 50, "LineStyle", "none")

            % plot walls, if asked

            if isfield(config, "plot_wall")==1 && config.plot_wall == 1
                hold on
                fig = obj.PlotWalls(equi,fig, config)
            end


            % plot magnetic flux lines, if asked

            if isfield(config, "psi_lines")==1
                hold on
                contour(equi.geo.grid.Rg, equi.geo.grid.Zg, equi.psi_n, config.psi_lines, '-w', 'LineWidth', 1.5)
            end
            
            %complete the plot

            uom = FieldUnitOfMeasurement(obj,field);
            title(field + uom)
            colorbar()
            colormap("jet")
            axis equal
            xlabel("R [m]")
            ylabel("Z [m]")

        end


        %% Plot Diagnostics

        function fig = PlotDiagnostics(obj,equi,diag,fig,config)


            if nargin < 2
                disp("missing input")
                return
            elseif nargin == 3
                fig = figure();
                config = struct;
            elseif nargin == 4
                config = struct;
            end

            %% Get the subplot information and set if not specified

            if isfield(config, "subplot")==0
                config.subplot = [1 1 1];
            end

            %% Plot

            figure(fig)
            subplot(config.subplot(1),config.subplot(2), config.subplot(3))
            
            if isfield(config, "hold") == 1 && config.hold == "on"
                hold on
            else
                hold off
            end

            % plot walls, if asked

            if isfield(config, "plot_wall")==1 && config.plot_wall == 1
                fig = obj.PlotWalls(equi,fig, config)
            end

            if isfield(config,"number_of_colours")
                num = config.number_of_colours;
            else
                num = 1;
                if isa(diag,"Diag_Bolo")
                    num = 7;
                end
            end

            SpecificColor = obj.TokaColor(diag,num);

            if isa(diag, "Diag_PickUpCoils")
                quiver(diag.R, diag.Z, diag.n(1,:), diag.n(3,:), 'color', SpecificColor,'LineWidth', 0.3,'ShowArrowHead','on','AutoScaleFactor',0.3)
                title("Pick-Up Coils")
                hold on
                plot(diag.R(1:end-5), diag.Z(1:end-5), '.', 'Color', SpecificColor, 'MarkerSize', 12)
                plot(diag.R(end-4:end), diag.Z(end-4:end), 'o', 'Color', SpecificColor, 'LineWidth',1, "HandleVisibility","off")
                plot(diag.R(end-4:end), diag.Z(end-4:end), '.', 'Color', SpecificColor, 'LineWidth',1, "HandleVisibility","off")

            elseif isa(diag, "Diag_FluxLoops")
                plot(diag.R, diag.Z, 's', 'LineWidth', 2, 'Color', SpecificColor)
                title("Flux Loops")

            elseif isa(diag, "Diag_SaddleCoils")
                plot([diag.R1; diag.R2], [diag.Z1; diag.Z2], '.-', 'LineWidth', 2, 'MarkerSize', 16, 'Color', SpecificColor)
                title("Saddle Loops")

            elseif isa(diag, "Diag_ThomsonScattering")
                plot(diag.R, diag.Z, '.', 'MarkerSize', 12, 'Color', 	SpecificColor)
                title("Thomson Scattering")

            elseif isa(diag, "Diag_InterferometerPolarimeter")
                plot([diag.R_in diag.R_out]', [diag.Z_in diag.Z_out]', '-', 'LineWidth', 2, 'Color', SpecificColor)
                title("Interferometer-Polarimeter")

            elseif isa(diag, "Diag_Bolo")

                [Z, ~, iZ] = unique(diag.Z_in);
                groups = arrayfun(@(k) find(iZ == k), 1:numel(Z), 'UniformOutput', false);

                for i = 1: length(groups)
                    plot([diag.R_in(groups{i}) diag.R_end(groups{i})]', [diag.Z_in(groups{i}) diag.Z_end(groups{i})]', '-', 'LineWidth', 1, 'MarkerSize', 16, 'Color', SpecificColor(i,:))
                    hold on
                end

                title("Bolometers")              
            end

            % complete the plot
            axis equal
            if isa(diag,"Diag_Bolo")==0
                xlim([equi.geo.R(1) equi.geo.R(end)]) 
                ylim([equi.geo.Z(1) equi.geo.Z(end)])
            else
                R_lim = [diag.R_in diag.R_end]; Z_lim = [diag.Z_in diag.Z_end];
                xlim([min(R_lim, [], "all") max(R_lim, [],  "all")])
                ylim([min(Z_lim, [],  "all") max(Z_lim, [],  "all")])
            end 
            xlabel("R [m]")
            ylabel("z [m]")
            colormap("jet")


        end

        %% Plot Measurements

        function fig = PlotMeasurements(~,diag,meas,fig,config)

            if nargin < 3
                disp("missing input")
                return
            elseif nargin == 3
                fig = figure();
                config = struct;
            elseif nargin == 4
                config = struct;
            end

            % Get the subplot information and set if not specified

            if isfield(config, "subplot")==0
                config.subplot = [1 1 1];
            end

            % Get the x axis label information

                R = 1 : length(diag.(meas));
                labelx = "ch";

            % Plot

            figure(fig)
            subplot(config.subplot(1),config.subplot(2), config.subplot(3))

            if isfield(config, "colour")==1
                SpecificColor = config.colour;
                titlep = "";
            else

                if isa(diag, "Diag_PickUpCoils")
                    titlep = "Pick-Up Coils"; % panel title
                    SpecificColor = "b"; % color plot
                    Unit = diag.unit; % unit of measurement for y axis label

                elseif isa(diag, "Diag_FluxLoops")
                    titlep = "Flux Loops";
                    SpecificColor = "#77AC30";
                    Unit = diag.unit;

                elseif isa(diag, "Diag_SaddleCoils")
                    titlep = "Saddle Loops";
                    SpecificColor = "r";
                    Unit = diag.unit;

                elseif isa(diag, "Diag_ThomsonScattering")
                    titlep = "Thomson Scattering " + meas;
                    SpecificColor = "#D95319";
                    Unit = diag.("unit_" + meas);

                elseif isa(diag, "Diag_InterferometerPolarimeter")
                    auxiliary_string_for_name = regexp(meas, '[A-Z]', 'match'); % procedure to extract the desiderd string
                    auxiliary_string_for_name = [auxiliary_string_for_name{:}];
                    if auxiliary_string_for_name(end) == "I"
                        auxiliary_string_for_name = auxiliary_string_for_name(1:end-1);
                    end
                
                titlep = auxiliary_string_for_name;
                SpecificColor = "#A2142F";
                Unit = diag.("unit_" + auxiliary_string_for_name);

                elseif isa(diag, "Diag_Bolo")
                titlep = "Bolometers";
                SpecificColor = "#A2142F";
                Unit = diag.unit;

                end
            end


            [R, R_sort] = sort(R);

            % two options available: errorbar or plot
            if isfield(config, "errorplot") == 1 && config.errorplot == 1
                errorbar(R, diag.(meas)(R_sort), diag.("sigma_" + meas)(R_sort), '.-','markersize', 10, 'LineWidth', 1, "Color", SpecificColor)
            else
                plot(R, diag.(meas)(R_sort), '.-', 'markersize', 16, 'LineWidth', 1.5, "Color", SpecificColor)
            end

            % complete the plot
            if isfield(config, "hold") == 1 && config.hold == "on"
                hold on
            else
                hold off
            end
            title(titlep)
            ylabel("[" + Unit +"]")
            xlabel(labelx)
            colormap("jet")

        end

        %% Function Plot Walls

        function fig = PlotWalls(obj,equi, fig, config)

            x1 = [equi.geo.R(1) equi.geo.R(end) equi.geo.R(end) equi.geo.R(1) equi.geo.R(1)];
            y1 = [equi.geo.Z(1) equi.geo.Z(1) equi.geo.Z(end) equi.geo.Z(end) equi.geo.Z(1)];

            x_extr = linspace(equi.geo.R(1),equi.geo.wall.R(1),100);
            y_extr = linspace(equi.geo.Z(1),equi.geo.wall.Z(1),100);

            %             ip_wall = inpolygon(x_extr, y_extr, equi.geo.wall.R, equi.geo.wall.Z);

            figure(fig)
            subplot(config.subplot(1),config.subplot(2),config.subplot(3))
            patch([x1 equi.geo.wall.R], [y1 equi.geo.wall.Z], [0.75 0.75 0.75], "HandleVisibility","off","edgecolor", "none")
            hold on
            plot(equi.geo.wall.R, equi.geo.wall.Z, '-k', 'LineWidth', 1.5)
            %             plot(x_extr(~ip_wall), y_extr(~ip_wall), 'color', [0.75 0.75 0.75], 'LineWidth', 2, "HandleVisibility","off")
            plot(equi.geo.R(1), equi.geo.wall.Z, '-k', 'LineWidth', 1, "HandleVisibility","off")
            %             plot(x_extr(ip_wall), y_extr(ip_wall), 'LineStyle', 'none')
            plot([equi.geo.wall.R(1) equi.geo.wall.R(end) equi.geo.wall.R(end) equi.geo.wall.R(1) equi.geo.wall.R(1)], [equi.geo.wall.Z(1) equi.geo.wall.Z(1) equi.geo.wall.Z(end) equi.geo.wall.Z(end) equi.geo.wall.Z(1)], '-k', 'LineWidth', 1, "HandleVisibility","off")

        end

        %% Function Field's unit of measurement

        function uom = FieldUnitOfMeasurement(obj,field)

            if field == "ne" || field == "ni"
                uom = " [m^{-3}]";
            elseif field == "Te" || field == "Ti"
                uom = " [eV]";
            elseif field == "psi"
                uom = " [Wb/rad]";
            elseif field == "Psi"
                uom = " [Wb]";
            elseif field == "Br" || field == "Bt" || field == "Bz"
                uom = " [T]";
            elseif field == "Jr" || field == "Jt" || field == "Jz"
                uom = " [A/m^{2}]";
            elseif field == "psi_n"
                uom = " [arb. units]";
            elseif field == "p"
                uom = " [Pa]";
            end
        end

        %% Function Specify Colors

        function SpecificColor = TokaColor(obj,diag,num)

            if isa(diag, "Diag_Bolo")
                SpecificColor = colormap(autumn(num));

            elseif isa(diag, "Diag_FluxLoops")
                SpecificColor = colormap(summer(num));

            elseif isa(diag, "Diag_SaddleCoils")
                SpecificColor = "r";

            elseif isa(diag, "Diag_ThomsonScattering")
                SpecificColor = colormap("lines");
                SpecificColor = SpecificColor(3,:);

            elseif isa(diag, "Diag_PickUpCoils")
                SpecificColor = "b";

            elseif isa(diag,"Diag_InterferometerPolarimeter")
                SpecificColor = colormap(pink(100));
                SpecificColor = SpecificColor(30,:);

            end


        end

    end
end

File: VirtualLab_Python/TokaPlot_Python/functions/TokaPlot.py
# -*- coding: utf-8 -*-
"""
Created on Thu Dec  4 16:41:24 2025

@author: novel
"""
import numpy as np
import os
import matplotlib.pyplot as plt
import matplotlib as mpl

from diagnostics.Tokalab.Diag_PickUpCoils import Diag_PickUpCoils
from diagnostics.Tokalab.Diag_SaddleCoils import Diag_SaddleCoils
from diagnostics.Tokalab.Diag_FluxLoops import Diag_FluxLoops
from diagnostics.Tokalab.Diag_ThomsonScattering import Diag_ThomsonScattering
from diagnostics.Tokalab.Diag_InterferometerPolarimeter import Diag_InterferometerPolarimeter

class TokaPlot:
    
    def __init__(self):
        self.field = None
        self.fig1 = None
        self.plot_colours = None
        self.uom = None
        # self.config = {}
            
    def plotfield(self,equi,field,fig,ax,config):
        
        R = equi.geo.R
        Z = equi.geo.Z
        
                
        if "subplot" not in config:
                config["subplot"] = [1, 1, 1]
       
        ax = fig.add_subplot(config["subplot"][0],config["subplot"][1],config["subplot"][2]+config["subplot"][3]+1)
        
        c = ax.contourf(equi.geo.R, equi.geo.Z, getattr(equi,field), 80, cmap='jet')
        if "plot_walls" in config and config["plot_walls"] == 1:
          
            self.plotwalls(ax,equi)
          
        if "psi_lines" in config:
            fig, ax.contour(equi.geo.R, equi.geo.Z, equi.psi_n, levels=config["psi_lines"], colors='w', linewidths=2)

             
        uom = self.fieldUOM(field)
        
        ax.set_title(field + uom)
        fig.colorbar(c, ax=ax, fraction=0.075, pad=0.075) #, ax=ax)
        ax.set_xlabel("R [m]")
        ax.set_ylabel("Z [m]")
        ax.set_aspect('equal')
        ax.set_xlim([min(R), max(R)])
        ax.set_ylim([min(Z), max(Z)])
        
        return ax
            
    def plotdiagnostics(self, equi, diag, fig, ax, config):               
            print("Sono dentro plot diagnostics")    
            
            R = equi.geo.R
            Z = equi.geo.Z
            
                    
            # if "subplot" not in config:
            #         config["subplot"] = [1, 3, 0, 2]
           
            # ax = fig.add_subplot(config["subplot"][0],config["subplot"][1],config["subplot"][2]+config["subplot"][3]+1)
            
        
            if "plot_walls" in config and config["plot_walls"] == 1:
              
              self.plotwalls(ax, equi)
              
                             
            if "n_ofcolours" in config:
                   n_ofcolours = config["n_ofcolours"]
            else: 
                if isinstance(diag, Diag_InterferometerPolarimeter)==1:
                 n_ofcolours = len(diag.R_in) + 1
                elif isinstance(diag, Diag_SaddleCoils):
                 n_ofcolours = len(diag.R1) + 1
                else:
                    n_ofcolours = len(diag.R) + 1
               
            # if isinstance(diag, Diag_Bolo):
            #   n_ofcolours = 7
             
            plot_colours = self.tokacolor(diag,n_ofcolours)
            
            if isinstance(diag, Diag_PickUpCoils):  
                ax.quiver(diag.R, diag.Z, diag.n[0, :], diag.n[2, :],color=plot_colours,linewidth=0.3, angles='xy', scale_units='xy', scale=0.95)
                ax.set_title("Pick-Up Coils")
                ax.plot(diag.R[:-5], diag.Z[:-5],'.', color=plot_colours, markersize=12)
                ax.plot(diag.R[-5:], diag.Z[-5:],'o', color=plot_colours, linewidth=1)
                ax.plot(diag.R[-5:], diag.Z[-5:],'.', color=plot_colours, linewidth=1)
                ax.legend().set_visible(False)
                ax.set_xlabel("R [m]")
                ax.set_ylabel("Z [m]")
                ax.set_aspect('equal')
                ax.set_xlim([min(R), max(R)])
                ax.set_ylim([min(Z), max(Z)])

            elif isinstance(diag, Diag_FluxLoops):
                ax.plot(diag.R, diag.Z, 's', color = plot_colours, linewidth=1, markersize=6)
                ax.set_title("Flux Loops")
                ax.set_xlabel("R [m]")
                ax.set_ylabel("Z [m]")
                ax.set_aspect('equal')
                ax.set_xlim([min(R), max(R)])
                ax.set_ylim([min(Z), max(Z)])
                
                
            elif isinstance(diag, Diag_SaddleCoils):
                ax.plot([np.array(diag.R1), np.array(diag.R2)], [np.array(diag.Z1), np.array(diag.Z2)], '.-', color = plot_colours, linewidth=1, markersize=6)
                ax.set_title("Flux Loops")
                ax.set_xlabel("R [m]")
                ax.set_ylabel("Z [m]")
                ax.set_aspect('equal')
                ax.set_xlim([min(R), max(R)])
                ax.set_ylim([min(Z), max(Z)]) 
                
            elif isinstance(diag, Diag_ThomsonScattering):
                 ax.plot(diag.R, diag.Z, '.', color = plot_colours, linewidth=1, markersize=3)
                 ax.set_title("Thomson Scattering")
                 ax.set_xlabel("R [m]")
                 ax.set_ylabel("Z [m]")
                 ax.set_aspect('equal')
                 ax.set_xlim([min(R), max(R)])
                 ax.set_ylim([min(Z), max(Z)]) 
                 
            elif isinstance(diag, Diag_InterferometerPolarimeter):
                 fig, ax.plot([np.array(diag.R_in), np.array(diag.R_out)], [np.array(diag.Z_in), np.array(diag.Z_out)], '-', color = plot_colours, linewidth=1)
                 ax.set_title("Interferometer Polarimeter")
                 ax.set_xlabel("R [m]")
                 ax.set_ylabel("Z [m]")
                 ax.set_aspect('equal')
                 ax.set_xlim([min(R), max(R)])
                 ax.set_ylim([min(Z), max(Z)])      

            # ax.plot(diag.R, diag.Z, color=plot_colours)
            return ax
            
    def plotmeasurements(self, diag, meas, fig, ax, config):

        
        y_value = getattr(diag,meas)
        
        R = np.linspace(1, y_value.size + 1, y_value.size, dtype=int)
        
        if "n_ofcolours" in config:
               n_ofcolours = config["n_ofcolours"]
        else: 
            if isinstance(diag, Diag_InterferometerPolarimeter)==1:
             n_ofcolours = len(diag.R_in) + 1
            elif isinstance(diag, Diag_SaddleCoils):
             n_ofcolours = len(diag.R1) + 1
            else:
                n_ofcolours = len(diag.R) + 1
                
        plot_colours = self.tokacolor(diag,n_ofcolours)
        uom = self.MeasUOM(diag, meas)
                
        if "errorbar" in config and config["errorbar"]==1:
            y_value=y_value.squeeze()
            y_err = np.ones((1,y_value.size))*getattr(diag, "sigma_" + meas)
            y_err = y_err.squeeze()
            ax.errorbar(R, y_value, y_err, fmt='-', marker='.', label=meas) 
        else: 
            ax.plot(getattr(diag, meas), '-', marker='.', label=meas)
        
        ax.set_xlabel("# of channel")
        ax.set_ylabel(uom)
        ax.legend()
        
        return ax
                 
    def plotwalls(self, ax, equi):
        
             ax.plot(equi.geo.wall.R, equi.geo.wall.Z, '-k', linewidth=2)
             ax.fill(
              [equi.geo.R[0], equi.geo.R[-1], equi.geo.R[-1], equi.geo.R[0], equi.geo.R[0]] + list(equi.geo.wall.R),
              [equi.geo.Z[-1], equi.geo.Z[-1], equi.geo.Z[0], equi.geo.Z[0], equi.geo.Z[-1]] + list(equi.geo.wall.Z),
              color=[0.75, 0.75, 0.75],
              label="_nolegend_")
             

    def tokacolor(self,diag,n_ofcolours):
        
        if isinstance(diag, Diag_PickUpCoils):
           plot_colours = "b"
           return plot_colours
        if isinstance(diag, Diag_FluxLoops):
           plot_colours = "forestgreen"
           return plot_colours
        if isinstance(diag, Diag_SaddleCoils):
           plot_colours = "r"
           return plot_colours
        if isinstance(diag, Diag_ThomsonScattering):
          plot_colours = "orangered"
          return plot_colours
        if isinstance(diag, Diag_InterferometerPolarimeter):
          plot_colours = "blueviolet"
          return plot_colours
        # if isinstance(diag, "Diag_Bolo"):
        #    plot_colours = "autumn" 
        #    return plot_colours
       
        
    def fieldUOM(self, field):
      
        if field == "ne" or field == "ni":
            uom = " $[m^{-3}]$"
            return uom
        elif field == "Te" or field == "Ti":
            uom = " [eV]"
            return uom
        elif field == "psi":
            uom = " [Wb/rad]"
            return uom
        elif field == "Psi":
            uom = " [Wb]"
            return uom
        elif field == "Br" or field == "Bt" or field == "Bz":
            uom = " [T]"
            return uom
        elif field == "Jr" or field == "Jt" or field == "Jz":
            uom = " $[A/m^{2}]$"
            return uom
        elif field == "p":
            uom = " [Pa]"
            return uom
        elif field == "psi_n":
            uom = " [arb. units]"           
            return uom
            
    def MeasUOM(self, diag, meas):
          
        if isinstance(diag, Diag_PickUpCoils)==1: 
            uom = " $[T]$"
            return uom
        elif isinstance(diag, Diag_FluxLoops)==1 or isinstance(diag, Diag_SaddleCoils)==1: 
            uom = " [Wb/rad]"
            return uom
        elif isinstance(diag, Diag_ThomsonScattering)==1 and (meas=="ne" or meas=="ni"):
            uom = " $[m^{-3}]$"
            return uom
        elif isinstance(diag, Diag_ThomsonScattering)==1 and (meas=="Te" or meas=="Ti"):
            uom = " [eV]"
            return uom
        elif isinstance(diag, Diag_InterferometerPolarimeter)==1 and (meas=="LIDh" or meas=="LIDc"):
            uom = " $[m^{-2}]$"
            return uom
        elif isinstance(diag, Diag_InterferometerPolarimeter)==1 and (meas=="FARc" or meas=="FARh" or meas=="FARc_typeI" or meas=="FARh_typeI"):
            uom = " [rad]"
            return uom
        elif isinstance(diag, Diag_InterferometerPolarimeter)==1 and (meas=="CMc" or meas=="CMh" or meas=="CMc_typeI" or meas=="CMh_typeI"):
            uom = " [rad]"           
            return uom
                